src/HttpLayer.hs

```diff
@@ -40,41 +42,62 @@ emptyHTTPNetworkLayer = HTTPNetworkLayer
 ------------------------------------------------------------
 
 -- | General api request function
-apiRequest :: Config -> Text -> Request
-apiRequest Config{..} u =
-    setRequestPath (encodeUtf8 path) $
-    addRequestHeader "Content-Type" "application/json" $
-    setRequestBasicAuth
-        (encodeUtf8 cfgEmail <> "/token")
-        (encodeUtf8 cfgToken) $
-    parseRequest_ (toString (cfgZendesk <> path))
+apiRequest
+    :: Config
+    -> Text
+    -> Either String Request
+apiRequest Config{..} u = mapLeft show $ catches buildRequest handlerList
   where
+    buildRequest :: forall m. MonadCatch m => m Request
+    buildRequest = do
+        req <- parseUrlThrow (toString (cfgZendesk <> path)) -- TODO(md): Get a list of exceptions that parseUrlThrow can throw
+        return $ setRequestPath (encodeUtf8 path) $
+                 addRequestHeader "Content-Type" "application/json" $
+                 setRequestBasicAuth
+                     (encodeUtf8 cfgEmail <> "/token")
+                     (encodeUtf8 cfgToken) req
+    handlerList :: MonadCatch m => [Handler m Request]
+    handlerList = [
+                    handlerJSON
+                  , handlerHTTP
+                  ]
+    handlerJSON :: MonadCatch m => Handler m Request
+    handlerJSON = Handler $ \(ex :: JSONException) -> error $ show ex
+    handlerHTTP :: MonadCatch m => Handler m Request
+    handlerHTTP = Handler $ \(ex :: HttpException) -> error $ show ex
     path :: Text
     path = "/api/v2" <> u
 
 -- | Api request but use absolute path
-apiRequestAbsolute :: Config -> Text -> Request
+apiRequestAbsolute
+    :: Config
+    -> Text
+    -> Either String Request
 apiRequestAbsolute Config{..} u =
-    addRequestHeader "Content-Type" "application/json" $
+    Right <$> addRequestHeader "Content-Type" "application/json" $
     setRequestBasicAuth (encodeUtf8 cfgEmail <> "/token") (encodeUtf8 cfgToken) $
     parseRequest_(toString u)
 
 -- | Request PUT
-addJsonBody :: forall a. ToJSON a => a -> Request -> Request
-addJsonBody body req = setRequestBodyJSON body $ setRequestMethod "PUT" req
+addJsonBody :: forall a. ToJSON a => a -> Request -> Either String Request
+addJsonBody body req = mapLeft show $ catch updateReq Left
+  where
+    updateReq :: Exception e => Either e Request
+    updateReq = Right <$> setRequestBodyJSON body $ setRequestMethod "PUT" req
+
 
 -- | Make an api call
 -- TODO(ks): Switch to @Either@.
 apiCall
     :: forall m a. (MonadIO m, FromJSON a)
     => (Value -> Parser a)
     -> Request
-    -> m a
+    -> m (Either String a)
 apiCall parser req = do
     putTextLn $ show req
     v <- getResponseBody <$> httpJSON req
     case parseEither parser v of
-        Right o -> pure o
+        Right o -> pure $ Right o
         Left e -> error $ "couldn't parse response "
```

Comment:
So a `Left` returns what? An error? I think your work here is done, I told you this is an _example_.

Reply:
I am sorry, I didn't understand what you said. Can you please rephrase?

Reply:
Seems like that matches the return type of the function (hint, hint) :smirk: